votes/0162-new-pr-requirement-and-commit-message-style.md

```diff
@@ -0,0 +1,166 @@
+
+## Solution
+
+The whole solution could be divided into two main parts.
+
+1. Refactor the pull request and issue association rule. 
+
+Currently the bot checks whether there is the required issue number reference in any of the commits. So to fulfill the requirement, contributors need to add the required issue information in at least one of the pull request commit messages. We'd like to drop this rule and apply a new one. 
+
+After this proposal, the bot will check whether there is relevant issue number(s) in the PR body, by `Issue Number: ref #{issue-number}` or `Issue Number: close[sd]?|resolve[sd]?|fix(e[sd])? #{issue-number}`. `ref` means associated with an issue but does not close it. There should be no other content in the same line and the check is case-insensitive. Multiple issues should be separated by a comma, like `Issue Number: close #12341, close #12342`. For PRs trying to close issues in a different repository, contributors need to first create an issue in the same repository and use this issue to track. If the pull request body does not provide the required content, the bot will add the `do-not-merge/needs-linked-issue` label to the pull request to prevent it from being merged.
+
+While merging the PR, the bot will extract the content after `Issue Number:` and add them as a new line to the final commit message body. So if a PR `#12345` has a title `pkg: what's changed in this one package` and its body contains `Issue Number: close #12341, close #12342`, after merging, its commit message is
+
+```
+pkg: what's changed in this one package (#12345)
+
+close #12341, close #12342
+```
+
+2. Refactor the final commit message style
+
+We hope to encourage contributors to write good, meaningful commit messages, drop useless messages, make the final commit message compact and clean. The current complex and messy merging rule will be changed to:
+
+The one-line subject rule will not be changed. It's still the pull request title. The final commit message body will be completely refactored. It won't take the commit messages of the commits in the pull request any more. We'll introduce a new element in the PR body:
+
+    ```commit-message
+    any multiple line commit messages that go into
+    the final commit message body
+
+    * fix something 1
+    * fix something 2
+    ``` 
+
+The content in the `commit-message` code block will be extracted and pasted into the final commit message body. The DCO requirement is still needed. But along with the `Co-authored-by` information, they will be grouped and distinguished by the bot and be appended to the final commit message. So after the proposal, for the first sample commit message in the Motivation section, it will be like:
+
+```
+rsmeter: add pubsub datasink  (#11719)
+
+Signed-off-by: Zhenchi <zhongzc_arch@outlook.com>
```

Comment:
Can bot extract `Signed-off-by` and `Co-authored-by` from original commit message automatically without asking  author to explicitly put them into code block?

Bot should be able to keep original order of sign off messages and eliminate duplicated sign offs automatically. This way authors are only required to write down whatever messages they really care about in code block.

Reply:
Yes, that's our goal. The bot will extract the `Signed-off-by` and `Co-authored-by` from the original commit messages and eliminate duplicates. The new code block should only contains any other meaningful descriptions explicitly written by the author. And I am not sure the order of sign off messages matters, but definitely not a hard work.

Reply:
The content filled out in the code block does not need to be included `Signed-off-by` and `Co-authored-by`.

Here is an actual example：https://github.com/ti-community-infra/test-dev/pull/497

Reply:
So we can remove signoff messages from code block in this example.

Reply:
This example is not the code block in the pull request body. It's the improved final commit message for the first sample in the Motivation section.

Reply:
Oops, I was reading original md and didn't render it in my mind😂